BUGFIX: #3513 #3441 meta keys "sparse" DataStructure

mhsdesign opened the pull request:
## This is the minefield PR for the range 5.3 - 7.3 (later stuff is fixed)

current state summed up here: https://github.com/neos/neos-development-collection/pull/3568#issuecomment-1140829355

--------


fixes: #3513
fixes: #3441
fixes: https://github.com/neos/neos-development-collection/pull/3568 #3576 

#### untypedChildKey with meta key is still now an untypedChildKey!

If a meta key was set on an 'untyped' child of a DataStructure, it was counted as typed and such rendered to null.

Btw, ANY meta key would let a key without object, value etc... count as typed. So this wouldnt work until now:

```
root = Neos.Fusion:DataStructure {
    untypedKey {
        renderMe = 'please'
        @somethingAnything = "foo"
    }
}
```

Added tests in relation to untyped child keys of DataStructures

#### Unsetting typedChildKey === untypedChildKey?
currently thats not working, whether this should work could be discussed, but i think it was meant this way?
https://github.com/mhsdesign/neos-development-collection/blob/777d11ff1402e9dd47501bb55c6717b6201dd63d/Neos.Fusion/Tests/Functional/FusionObjects/Fixtures/Fusion/DataStructure.fusion#L146
https://github.com/mhsdesign/neos-development-collection/blob/777d11ff1402e9dd47501bb55c6717b6201dd63d/Neos.Fusion/Tests/Functional/FusionObjects/DataStructureTest.php#L142

**Checklist**
- [x] Code follows the PSR-2 coding style
- [x] Tests have been created, run and adjusted as needed
- [x] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)

mhsdesign committed:
BUGFIX: #3513 #3441 meta keys "sparse" DataStructure

If a meta key was set on an 'untyped' child of a DataStructure, it was counted as typed and such rendered to null.

Added tests in relation to untyped child keys of DataStructures

mhsdesign commented:
Looks like some unit tests dont like what i have done ... (since the meta key wont stop the datastructure implementation now from creating nested datastructures ...)

---------

i looked at the original pr https://github.com/neos/neos-development-collection/pull/2729 ...

maybe we should fix it like https://github.com/neos/neos-development-collection/pull/2729/files/ff31748ec26922d745a099f39baed7b30e9ef85c#r352892386
As there is really lot of internal stuff exposed.

and this sounds fun too ^^ https://github.com/neos/neos-development-collection/pull/2729#issuecomment-561039907

mhsdesign commented:
upmerging this will be fun since there where some changes and new things added to the (yet to fix) DataStructureImplementationTest along the road from 5.3 to 7.3 ...

mhsdesign committed:
FIX: Unit Test of Fusion DataStructureImplementation

Since __meta keys won't tell the DataStructureImplementation now anymore that it is untyped the tests need to be adjusted.

I used as base how DataStructureImplementationTest is written in Neos 7.3, since i did all the dev work on 7.3(master) and cherry-picked it.
Maybe that will make upmerging easier.

mhsdesign commented:
The currently failing `ArrayImplementationTest `which will become later the `JoinImplementationTest `made me realize, since `JoinImplementation `inherits from `DatastructureImplementation `it indirectly supports the feature of auto creating datastructures (https://github.com/neos/neos-development-collection/pull/2729)

is this wanted behavior? @mficzel 

see:
https://fusionpen.punkt.de/fusionpen/3807d45e274219afe23973fd4453f783c9d2eaa6.html

mficzel commented:
That is an accident … imho it would make sense for join to assume untyped keys to be join aswell

mhsdesign commented:
Hmm i dont think untyped nested Joins make sense ... but when it helps some...

The real problem i see, is that were using inheritance everywhere when building the php implementation... couldnt that better resolved with traits? that way we wont run into such bugs...

mficzel commented:
That inheritence mainly happened without much intention. Refactoring to traits would be fine, however i would also like to see a concept of defaultType for untyped nodes in fusion as we now have implemented this independently in different places already (offcourse this could also be a special trait):

Examples i am aware of:
- Neos.Fusion:DataStructure
- Neos.Fusion:Case 
- Neos.Fusion.Form. .. MultiStepProcess, SchemaCollection, ActionCollection

mhsdesign committed:
BUGFIX: #3576 Inherited implementations of DataStructureImplementation automatically create DataStructures

This bugfix is just a dirty fix. The code should be refactored: #3577

Since JoinImplementation inherits from DatastructureImplementation it indirectly supports the feature of auto creating Datastructures (#2729)

Neos.Fusion:Join for example. Or the deprecated Neos.Fusion:RawArray

mhsdesign committed:
BUGFIX: #3576 Inherited implementations of DataStructureImplementation automatically create DataStructures

Second more stable attempt.
The previous commit SHA: e974c6964eb690148c9fb2630e4a5685a65df9c7
Didn't account for when using the implementation class of DataStructureImplementation on a new fusion object like:

prototype(MyStuff:Data).@class = '...DataStructureImplementation'

Also, Fusion object inheritance from DataStructureImplementation was broken I think.

mhsdesign commented:
btw when you extend the DataStructure in fusion, woudnt it make sense that the nested untyped keys become automatically `Your:DataStructure`? (via `$this->fusionObjectName`)

`prototype(Your:DataStructure) < prototype(Neos.Fusion: DataStructure)
`
```
foo = Your:DataStructure {
    foo {
        hi = 123
    }
}
```

```
foo = Your:DataStructure {
    foo = Your:DataStructure {
        hi = 123
    }
}
```

mhsdesign commented:
Thinking more about it i think its cleaner to use `$this->fusionObjectName` instead of the hardcoded name:

https://github.com/neos/neos-development-collection/blob/78bcf8668e4a79f5e85f7771716ab9f2ee7546c8/Neos.Fusion/Classes/FusionObjects/DataStructureImplementation.php#L42

 since technically the fusion object could be named `prototype(My.Fusion:DataStructure).@class= "...DataStructureImplementation"` and Neos.Fusion:DataStructure could even be not available.

mhsdesign commented:
simply checking for the existance of any reversed parser parser path keys is just wrong. See __meta or even __prototypes (partial override)

the fusion parser should expose a list of parse keys which would identify the path as beeing used with vale/object/expression like:

`ParserInterface::ASSIGNED_PATH_KEYS`

mhsdesign commented:
i think the parser should not only expose a list of reserved parse tree keys, but an additional list with keys which would identify, that a path has an object or value etc... or since those keys should better not be too much api - create a method, to identify if a path has an object or value/eel assigned or not...

mhsdesign commented on Neos.Fusion/Classes/FusionObjects/DataStructureImplementation.php:
```diff
@@ -38,7 +39,8 @@ public function evaluate()
+            if ($this->isImplementationDataStructureImplementation()
+                && $this->isUntypedProperty($this->properties[$key])) {
                 $propertyPath .= '<Neos.Fusion:DataStructure>';
             }
```
i agree on the `'<' . $this->fusionObjectName . ' >';` part  (see https://github.com/neos/neos-development-collection/pull/3568#issuecomment-1018747823)

the `$this->isImplementationDataStructureImplementation()` barrier is to limit the untyped feature only to datastructures - otherwise with all this inheritance stuff we might never know what we allow with this unwillingly: see  https://github.com/neos/neos-development-collection/issues/3576

the ugly `isImplementationDataStructureImplementation` can be removed once we refactor #3577

mficzel replied:
How about having a defaultType configured in the data stucture implementation. 
That could be altered in derived implementations easyly 

```
class DataStructureImplementation extends AbstractArrayFusionObject
    protected $defaultFusionObjectType = 'Neos.Fusion:DataStructure';

    # used like:
    
    if ($this->isUntypedProperty($this->properties[$key]) &&  $this->defaultFusionObjectType) {
        $propertyPath .= '<' . $this->defaultFusionObjectType . ' >';
    }
```

```
class JoinImplementation extends DataStructureImplementation {
    protected $defaultFusionObjectType = 'Neos.Fusion:Join';
```

```
class HttpResponseImplementation extends DataStructureImplementation {
    protected $defaultFusionObjectType = null;
```

kdambekalns replied:
Sounds like a nice idea.

mhsdesign commented on Neos.Fusion/Classes/FusionObjects/DataStructureImplementation.php:
```diff
@@ -96,6 +98,21 @@ private function isUntypedProperty($property): bool
         }
-        return array_intersect_key(array_flip(Parser::$reservedParseTreeKeys), $property) === [];
+
+        $keysToIdentifyTypedProperty = array_flip(Parser::$reservedParseTreeKeys);
```
`$reservedParseTreeKeys = ['__meta', '__prototypes', '__stopInheritanceChain', '__prototypeObjectName', '__prototypeChain', '__value', '__objectType', '__eelExpression'];`

has to much stuff - for this feature we only need to check for:
one of these: `['__value', '__objectType', '__eelExpression']` - otherwise we are adding buggy functionality as we have see this issue ^^

should these 3 'magic' keys be inline in the `DataStructureImplementation` or be exposed seperately by the parser?

mficzel replied:
I would hardcode those now in the DataStructureImplementation.

mficzel replied:
BTW: Found this in fusion forms https://github.com/neos/fusion-form/blob/master/Classes/Runtime/FusionObjects/AbstractCollectionFusionObject.php#L130

mhsdesign commented on Neos.Fusion/Tests/Unit/FusionObjects/DataStructureImplementationTest.php:
```diff
@@ -14,118 +14,113 @@
                 'Position end should put element to end',
-                ['second' => ['__meta' => ['position' => 'end']], 'first' => ['__meta' => []]],
-                ['/first', '/second']
+                ['second' => ['__meta' => ['position' => 'end']], 'first' => []],
```
before this fix having a `__meta` key like `@if` or `@position` would lead to the key `second` not being identified as untyped.

now it is and we need to adjust these test.

mhsdesign commented on Neos.Fusion/Classes/FusionObjects/DataStructureImplementation.php:
```diff
@@ -96,6 +98,21 @@ private function isUntypedProperty($property): bool
         if (!is_array($property)) {
```
by making `private function isUntypedProperty(`

protected - we could override it in subclasses eg. in Join and always return false there ...

kdambekalns commented on Neos.Fusion/Classes/FusionObjects/DataStructureImplementation.php:
```diff
@@ -96,6 +98,21 @@ private function isUntypedProperty($property): bool
+     * (fx. returns false, when this class was further inherited.)
+     *
+     */
+    private function isImplementationDataStructureImplementation(): bool
```
1. The name should rather be `isImplementationADirectDataStructureImplementation()`
2. Is this still needed? The other mentioned related issues seem solved…

kdambekalns reviewed (commented):
Left two comments and would ask @mhsdesign and @mficzel to clarify the status of the open discussions.

Also, @mhsdesign, could you please rewrite the PR description to be clear about what this does, the consequences of the change? Also could you add some notes about what needs to be kept in mind when upmerging this?

Thanks!

mhsdesign commented:
@kdambekalns thanks for reviewing this ;)

this pr targets only neos 5.3 - 7.3

with 8.0 we have this bugfix/complete cleanup from @mficzel: https://github.com/neos/neos-development-collection/pull/3645

and nothing of this pr should land in the 8.0 branch or interfere with it...

i dont really know how to proceed.

the mentioned pr also took concerning some minor edge details a better road. So if we want to fix this bug for 5.3 - 7.3 i would adjust this pr a bit.

----

if you like we can chat about this and the whole issue sometime ;)

mhsdesign commented:
Ill close this as won't fix for 7.3 ;) Too sad since we put a lot of effort into it but the problem is just overly complex and we can be lucky to have it fixed with 8.0!
